data/transactions/logic/frames.go

```diff
@@ -70,14 +71,15 @@ func opFrameBury(cx *EvalContext) error {
 	}
 
 	frame := cx.callstack[top]
-	if i < 0 && -int(i) > frame.args {
+	// If proto was used, don't allow `frame_bury` to go below specified args
+	if frame.clear && -int(i) > frame.args {
```

Comment:
This makes a lot more sense to me now. But it does leave me wondering if `frame_dig/bury` should even be allowed when you're not "in a frame". I'm taking the most convservative definition that "being in a frame" is signaled by a `proto` call for the subroutine.

Reply:
I'll give the reasoning and what I think are the weaknesses in the reasoning.

The idea is that `proto` is optional, and intended to make a certain kind of subroutine easy to write: One that does not remove its (known number of) arguments from the stack, places its (known number of) return values on top of them, potentially uses more stack for locals, and then retsubs. If you are writing a subroutine that *doesn't* follow that script, you can't use `proto`, but you might still take advantage of `frame_{dig,bury}` as a convenience for accessing your arguments and placing return values.  As an example from yesterday, suppose you write `split`, and it will break up a []byte argument into a bunch of pieces, perhaps using `""` as a terminator on the stack, or perhaps also returning the number of pieces as an explicit return value (so the caller knows how much to pop).

Such a function might want to use `frame_{dig,bury}`, but can't use `proto`

The weakness of this argument is that most such a functions would have some trouble using them. They only work well if the function is willing to abide by the rule "don't actually remove your arguments from the stack, so frame_get still sees args, retvals, and locals at the expected places"  And it would have a hard time using `frame_put`, since `frame_put` takes static immediate values, but you're trying to place an unknown number of retvals. And then you probably wish you could remove the arguments when you're finally done (as `retsub` would do it you could have used `proto`) but there is no opcode to do that.  It's possible that adding a mechanism to manipulate the fp, and/or to perform that final "slide" would make things more usable for these types of functions.

Reply:
Perhaps a more complex version of `retsub` that only these sorts of apps would need to use. It would, in essence, take the `a` and `n` as stack arguments so that it can perform the cleanup that `retsub` would have done if `proto` had been used. The basic idea is that if you know your function is simple and has a static number of args/returns, you use `proto`, if not, and you still want to use the `frame_` and "retsub cleanup" conveniences, you explain how many args and retvals you had at the _end_, once you've figured it out.